[Merged by Bors] - Doppelganger detection

realbigsean opened the pull request:
## Issue Addressed

Resolves #2069 

## Proposed Changes

- Adds a `--doppelganger-detection` flag
- Adds a `lighthouse/seen_validators` endpoint, which will make it so the lighthouse VC is not interopable with other client beacon nodes if the `--doppelganger-detection` flag is used, but hopefully this will become standardized. Relevant Eth2 API repo issue: https://github.com/ethereum/eth2.0-APIs/issues/64
- If the `--doppelganger-detection` flag is used, the VC will wait until the beacon node is synced, and then wait an additional 2 epochs. The reason for this is to make sure the beacon node is able to subscribe to the subnets our validators should be attesting on. I think an alternative would be to have the beacon node subscribe to all subnets for 2+ epochs on startup by default.

## Additional Info

I'd like to add tests and would appreciate feedback. 

TODO:  handle validators started via the API, potentially make this default behavior

realbigsean committed:
add doppelganger endpoint

realbigsean committed:
add doppelganger detection methods and fix endpoint

realbigsean committed:
add validator client logic and API test

realbigsean committed:
add simulator test and add log

realbigsean committed:
add log

realbigsean committed:
fix typo and comparator bug

realbigsean committed:
move log

realbigsean committed:
delete log

realbigsean committed:
fix doppelganger detection loop

realbigsean committed:
add breaks

torfbolt commented:
For reference and comparison, nimbus has already implemented this feature in https://github.com/status-im/nimbus-eth2/pull/2244

paulhauner commented on validator_client/src/lib.rs:
```diff
@@ -291,21 +299,55 @@ impl<T: EthSpec> ProductionValidatorClient<T> {
+        // Configure doppelganger detection if enabled.
+        let (block_service, attestation_service) = if config.doppelganger_detection {
+            // block until node is synced
+            tokio::select! {
```
Seems like an unnecessary `tokio::select`.

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2356,6 +2356,41 @@ pub fn serve<T: BeaconChainTypes>(
+            |query: api_types::SeenValidatorQuery, chain: Arc<BeaconChain<T>>| {
+                blocking_json_task(move || {
+                    let indices: Vec<usize> = chain
+                        .head_beacon_state()
```
We could avoid cloning a `BeaconState` here and use the `validator_pubkey_cache`:

https://github.com/sigp/lighthouse/blob/46920a84e857d6c086adf4b42810ac03de94ed3e/beacon_node/beacon_chain/src/beacon_chain.rs#L728-L747

realbigsean replied:
edit: moved this to a separate comment. Accidentally left a reply in the wrong spot 😆

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2356,6 +2356,41 @@ pub fn serve<T: BeaconChainTypes>(
+                        })
+                        .collect();
+
+                    let seen_validator =
```
Maybe it would be nice to return a list of the offending validator indices to make it easier for the user to determine which validator is duplicated.

michaelsproul committed:
Use the database to persist the pubkey cache

paulhauner reviewed (commented):
Thanks for getting started on this one! This is a bit of a fly-by review aimed at giving some initial feedback, I suspect that's what you're after :)

I made a few comments, but there's a couple of other things I wanted to touch on too.

1. We will need to consider how this performs prior to genesis. I think that if we start the VC prior to genesis we want to skip DGP (doppelganger protection) completely. Otherwise, we'll end up with no nodes producing blocks for the first couple of epochs.
1. I think we might want to turn on DGP protection by default, since it's the safest route. Users can opt into fast and risky if they want.
1. I wonder if this will falsely detect a doppelganger if we reboot in the middle of the epoch. E.g., if the validator attested in the first slot, we reboot in the second slot and then we check for DGs in the third slot and then get a hit because we find our validator in the same epoch.
1. What are your thoughts on writing DGP as a separate routine which we run before starting the block/attestation/duties services? It might turn out a little simpler to implement and perhaps less likely for us to forget to gate something behind it in the future. I haven't thought deeply about this, just raising it :)

realbigsean committed:
pr updates

realbigsean committed:
pr updates

paulhauner committed:
Add first pass at using pre-state for block prod

paulhauner committed:
Add warnings for late block broadcasts

paulhauner committed:
Move warnings for late blocks

paulhauner committed:
Add comments

paulhauner committed:
Use crit for late blocks

paulhauner committed:
Tidy docs, add sanity check

paulhauner committed:
Add more metrics for duties

paulhauner committed:
Refactor proposer duties endpoint

paulhauner committed:
Fix bug in state advance timer

paulhauner committed:
Remove old beacon proposer cache

paulhauner committed:
First refactor of attester duties

paulhauner committed:
Merge branch 'pr/2234' into faster-duties

paulhauner committed:
Add pubkey bytes struct

paulhauner committed:
Almost complete duties refactor

paulhauner committed:
Add almost-compiling refactor

paulhauner committed:
Remove old duties service code

paulhauner committed:
Merge branch 'unstable' into faster-duties

paulhauner committed:
Fix compile errors

paulhauner committed:
Remove old validator_duty struct

paulhauner committed:
Fix attestation, indices bugs

paulhauner committed:
Fix clippy lints, add metrics

paulhauner committed:
Specify integer types

paulhauner committed:
Revert "Specify integer types"

This reverts commit 60e1a7dbc73db1b9b8870b1e449ab0c1cc39b881.

paulhauner committed:
Only subscribe for later slots

paulhauner committed:
Specify integers as slots

paulhauner committed:
Use saturating math

paulhauner committed:
Subscribe even later

realbigsean committed:
make doppelganger detection default behavior, move it to a service, track per validator

realbigsean committed:
Merge branch 'faster-duties' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	validator_client/src/block_service.rs
	validator_client/src/duties_service.rs
	validator_client/src/initialized_validators.rs
	validator_client/src/lib.rs

paulhauner committed:
Fix slashing test compile errors

realbigsean committed:
clean up

realbigsean committed:
clippy fixes

paulhauner committed:
Fix slashing db compile error

paulhauner committed:
Add comments, fancy block retry logic

paulhauner committed:
Fix test compile error

paulhauner committed:
Tidy

paulhauner committed:
Break proposer duties into own file, refactor

realbigsean committed:
update error handling and filtering for signing keys

realbigsean committed:
Merge branch 'faster-duties' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	validator_client/src/duties_service.rs

realbigsean committed:
return which validators have been seen from the `seen_validators` endpoint

paulhauner committed:
Avoid bouncing pubkey lock

paulhauner committed:
Improve test coverage, fix bugs

paulhauner committed:
Test for proposer cache changes

paulhauner committed:
Test proposer duties from future epoch

paulhauner committed:
Add first pass of attn duties refactor

paulhauner committed:
Fix failing test

paulhauner committed:
Cache next-epoch duties during state advance

paulhauner committed:
Update comment

paulhauner committed:
Add more debugging for shuffling cache

paulhauner committed:
Reduce logging for cache

paulhauner committed:
Add more VC metrics

paulhauner committed:
Avoid state read for committee cache

paulhauner committed:
More VC metrics

paulhauner committed:
Fix clippy lint

realbigsean committed:
Merge branch 'faster-duties' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
tests

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/beacon_chain/src/state_advance_timer.rs
	beacon_node/http_api/src/attester_duties.rs
	beacon_node/http_api/src/proposer_duties.rs
	consensus/types/src/beacon_state.rs
	validator_client/src/duties_service.rs
	validator_client/src/initialized_validators.rs
	validator_client/src/validator_store.rs

realbigsean committed:
merge fixes

realbigsean committed:
fix sim

realbigsean committed:
debugging

realbigsean committed:
fix slot clock read

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	testing/simulator/src/no_eth1_sim.rs

realbigsean committed:
fix no eth1 sim

realbigsean committed:
cargo fmt

realbigsean committed:
fix logic to keep validators from proposing/attesting

realbigsean committed:
update log

realbigsean committed:
update log

realbigsean committed:
critical bug

realbigsean committed:
move indices to validator store

realbigsean committed:
move indices to validator store

realbigsean committed:
fix log

realbigsean committed:
fix log

realbigsean committed:
shutdown sender testing

realbigsean committed:
attempt to fix simulator

realbigsean committed:
attempt to fix simulator

realbigsean committed:
revert simulator changes

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	validator_client/src/initialized_validators.rs

realbigsean commented:
I think this is in pretty good shape at the moment. I spent a decent amount of time trying to fit test scenarios in the simulator, but it's turning out to not be the best place for it because all validator clients share the same `Environment` which makes it difficult to test a single validator client exiting.  I'm gonna instead try to set something up with local testnet scripts so this is a bit blocked on #2229 getting merged

paulhauner commented:
> this is a bit blocked on #2229 getting merged

I've just triggered bors for that PR :)

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
doppelganger detection test script

realbigsean committed:
fix timeout

realbigsean committed:
debug

realbigsean committed:
move to script test dir in test suite

realbigsean committed:
test failure

realbigsean committed:
test failure

realbigsean committed:
test success

realbigsean committed:
Accept validator ids rather than pubkeys in the seen_validators endpoint. Remove pubkeys from the doppelganger detection log. Update github actions step name.

realbigsean commented:
Made a bunch of updates here, and ready for another round of review. To address some of the questions posed earlier @paulhauner 

> 1. We will need to consider how this performs prior to genesis. I think that if we start the VC prior to genesis we want to skip DGP (doppelganger protection) completely. Otherwise, we'll end up with no nodes producing blocks for the first couple of epochs.

I've made it so if doppelganger detection is enabled for a validator at or before the genesis epoch, nothing happens (validator runs normally and doesn't check for doppelgangers)

> 2. I think we might want to turn on DGP protection by default, since it's the safest route. Users can opt into fast and risky if they want.

I've set it to run on default and added an opt out flag. I do think that there's a risk we will have a lot of people upset about missing ~3 attestations per validator on restart if we aren't clear enough about the change though. Also, do you think this presents any risk around lighthouse releases (maybe an emergency hotfix or something)? if some large portion of the network updates at once?

> 3. I wonder if this will falsely detect a doppelganger if we reboot in the middle of the epoch. E.g., if the validator attested in the first slot, we reboot in the second slot and then we check for DGs in the third slot and then get a hit because we find our validator in the same epoch.

I've made it so whenever the validator starts, it detects attestations in (current epoch + 1) and (current epoch + 2), until the end of (current epoch + 2). So I don't think we will detect anything on reboots, unless there is an edge case in there around clock drift?

> 4. What are your thoughts on writing DGP as a separate routine which we run before starting the block/attestation/duties services? It might turn out a little simpler to implement and perhaps less likely for us to forget to gate something behind it in the future. I haven't thought deeply about this, just raising it :)

Separated into a separate service :)

realbigsean committed:
increase doppelganger detection timeout time

realbigsean committed:
typo

paulhauner commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2775,6 +2775,25 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        validator_indices: &[usize],
+        epoch: &Epoch,
+    ) -> impl Iterator<Item = usize> + '_ {
+        self.observed_attesters
```
It's probably a bit pedantic, but if it's simple enough perhaps we should check for aggregators and block producers too?

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/beacon_node/beacon_chain/src/beacon_chain.rs#L217

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/beacon_node/beacon_chain/src/beacon_chain.rs#L219

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2127,6 +2127,45 @@ pub fn serve<T: BeaconChainTypes>(
         });
 
+    // GET lighthouse/seen_validators?ids,epochs
+    let get_lighthouse_seen_validators = warp::path("lighthouse")
```
I made a PR on the APIs repo to try and make a standard endpoint: https://github.com/ethereum/eth2.0-APIs/pull/131. I figure that if we want our VC to support eth2 Infura then we'll need to get this endpoint into the spec. I modified the request/response a bit to try and make it more palatable to the spec repo.

So, there's three scenarios with the spec endpoint:

1. It is accepted very quickly and merged in the next couple of days (probably unlikely)
2. It is accepted in the coming weeks (seems most likely, I hope)
3. It is rejected.

If (1) happens, then we can just update this PR to use the spec endpoint. Otherwise, perhaps we should just modify this endpoint to look roughly like what the spec one will look like (i.e., my extra fields) and then we can just leave it as a `lighthouse/` endpoint. Then, once the spec merges we can add that new endpoint to the BN but leave the old `lighthouse/` endpoint hanging around throwing deprecated WARNs for a few versions before killing it. That'll give a bit of a grace period to users.

What are your thoughts about all of this?

realbigsean replied:
Agree that it'll probably be number 2. Maybe it'll be merged along with altair related changes? Regardless, I implemented your proposal but on a `lighthouse/` endpoint

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2127,6 +2127,45 @@ pub fn serve<T: BeaconChainTypes>(
+                                .map_err(warp_utils::reject::beacon_chain_error)?
+                                .ok_or_else(|| {
+                                    warp_utils::reject::custom_not_found(format!(
+                                        "Could not find index for validator public key: {}",
```
```suggestion
                                        "could not find index for validator public key: {}",
```

Potentially more consistent with other server errors.

paulhauner commented on beacon_node/http_api/tests/tests.rs:
```diff
@@ -2083,6 +2083,49 @@ impl ApiTester {
+            .unwrap()
+            .data;
+
+        assert!(!result.is_empty());
```
Should we perhaps check that we've seen all these validators? I can't recall what state they'll be in after the test harness does its stuff, but it *should* be deterministic. Not a biggy.

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2127,6 +2127,45 @@ pub fn serve<T: BeaconChainTypes>(
+        .and(warp::query::<api_types::SeenValidatorQuery>())
+        .and(chain_filter.clone())
+        .and_then(
+            |query: api_types::SeenValidatorQuery, chain: Arc<BeaconChain<T>>| {
```
Having the list of indices in the query path might be an issue for our Pyrmont/Prater nodes with 2.5k validators. I think it'll result in a URL that's too-long.

This would be resolved by adopting the `POST` approach in https://github.com/ethereum/eth2.0-APIs/pull/131.

realbigsean replied:
Very good point!

paulhauner commented on validator_client/src/cli.rs:
```diff
@@ -181,4 +181,18 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
+                .long("disable-doppelganger-detection")
+                .value_name("DISABLE_DOPPELGANGER_DETECTION")
+                .help("By default, Lighthouse will delay startup for two epochs and monitor for \
+                    attestations on the network by any of the validators managed by this client. \
```
```suggestion
                    messages on the network by any of the validators managed by this client. \
```

Only if we decided to check blocks and aggregates.

paulhauner commented on validator_client/src/cli.rs:
```diff
@@ -181,4 +181,18 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
+                    that you are running a second validator client with the same keys. This \
+                    validator client will immediately shutdown if this is detected in order to \
+                    avoid potentially committing a slashable offense. Use this flag in order to \
+                    DISABLE this functionality, and begin attesting immediately.")
```
Could we maybe just add `USE WITH CAUTION` to the end of this too? :sweat_smile:

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,159 @@
+        info!(
+            log,
+            "Doppelganger detection service started";
+            "next_update_millis" => duration_to_next_slot.as_millis()
```
```suggestion
            "next_update" => ?duration_to_next_slot
```

I recently found out that this provides a pretty nice output. Might be a little nicer for users?

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,159 @@
+            .write()
+            .update_all_doppelganger_detection_epochs(current_epoch, genesis_epoch);
+
+        let mut interval = {
```
Pawan and I recently switched the other time-based loops in the VC over to a new format like this:

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/validator_client/src/duties_service.rs#L215-L238

The benefit of the new approach is that it will adjust itself if the system clock changes whereas the interval-based approach will ignore such changes and remain on the old schedule. It's also kinda simpler to define and it'll remove that `interval_at` panic :rocket:

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,159 @@
+        let slot = self.slot_clock.now().ok_or("Unable to read slot clock")?;
+        let epoch = slot.epoch(E::slots_per_epoch());
+
+        // get all validators in the doppelganger detection epoch
```
```suggestion
        // Get all validators requiring a check in this epoch.
```

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,159 @@
+                Err(e) => {
+                    warn!(
+                        log,
+                        "Failed complete query for doppelganger detection... Restarting doppelganger detection process.";
```
```suggestion
                        "Failed to complete query for doppelganger detection... Restarting doppelganger detection process.";
```

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,159 @@
+                    }
+                }
+                Err(e) => {
+                    warn!(
```
```suggestion
                    crit!(
```

Might be nice to make this loud since it can prevent validators from running.

paulhauner commented on validator_client/src/duties_service.rs:
```diff
@@ -114,7 +110,7 @@ pub struct DutiesService<T, E: EthSpec> {
     /// Returns the total number of validators known to the duties service.
     pub fn total_validator_count(&self) -> usize {
-        self.validator_store.num_voting_validators()
+        self.validator_store.num_enabled_validators()
```
This seems like a bit of a risky semantics change, would you mind explaining for me?

realbigsean replied:
Was thinking that doppelganger detection introduces periods where the validators returned by this method aren't currently voting. But you could definitely argue that all the validators returned by this method will either eventually be voting or will have a doppelganger out there voting for them. So I reverted it :slightly_smiling_face:

paulhauner commented on validator_client/src/duties_service.rs:
```diff
@@ -141,13 +137,19 @@ impl<T: SlotClock + 'static, E: EthSpec> DutiesService<T, E> {
-                    .filter(|proposer_data| proposer_data.slot == slot)
+                    .filter(|proposer_data| {
+                        proposer_data.slot == slot
+                            && signing_pubkeys.contains(&proposer_data.pubkey)
```
Perhaps we could make a `ValidatorStore::signing_pubkeys_hashset` which collects into a `HashSet` instead of `Vec` to avoid making this function `O(n^2)`. It might hurt a little on our testnet machines with 2.5k vals.

paulhauner commented on validator_client/src/duties_service.rs:
```diff
@@ -158,12 +160,18 @@ impl<T: SlotClock + 'static, E: EthSpec> DutiesService<T, E> {
-            .filter(|duty_and_proof| duty_and_proof.duty.slot == slot)
+            .filter(|duty_and_proof| {
+                duty_and_proof.duty.slot == slot
+                    && signing_pubkeys.contains(&duty_and_proof.duty.pubkey)
```
As per previous comment.

paulhauner commented on validator_client/src/http_api/mod.rs:
```diff
@@ -442,12 +457,27 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
                         ))),
                         Some(enabled) if enabled == body.enabled => Ok(()),
                         Some(_) => {
+                            let current_epoch = validator_store
```
Maybe a little helper function would be nice to reduce this duplication?

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -87,6 +95,9 @@ pub enum SigningMethod {
     signing_method: SigningMethod,
     graffiti: Option<Graffiti>,
+    doppelganger_detection_epoch: Option<Epoch>,
+    /// There is a task which ensures this value is kept up-to-date.
```
```suggestion
    /// The validators index in `state.validators`, to be updated by an external service.
```

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -611,4 +646,85 @@ impl InitializedValidators {
+    }
+
+    /// Updates the doppelganger detection epoch of all initialized validators based on the provided
+    /// slot.
```
```suggestion
    /// epoch.
```

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -611,4 +646,85 @@ impl InitializedValidators {
+    ) {
+        let detection_epoch =
+            self.determine_doppelganger_detection_epoch(current_epoch, genesis_epoch);
+        info!(self.log, "Monitoring for doppelgangers for all validators"; "detecting_until" => ?detection_epoch);
```
Is this log still relevant?

realbigsean replied:
Removed it, don't think it's necessary

paulhauner commented on validator_client/src/lib.rs:
```diff
@@ -142,15 +145,63 @@ impl<T: EthSpec> ProductionValidatorClient<T> {
+
+        // Perform some potentially long-running initialization tasks.
+        let (genesis_time, genesis_validators_root, fork) = tokio::select! {
+            tuple = init_from_beacon_node(&beacon_nodes, &context) => tuple?,
```
If we init from the beacon node *before* we initialize validators then we have this issue where we don't start decoding keypairs until after genesis.

This has burned me with some testnets where we have lots of validators, we miss the first few epochs cause we're decrypting keypairs.

I'm running out of time to figure out a solution right now, but perhaps we can find something?

realbigsean replied:
Oh I didn't realize this would happen.. I've updated it so that we don't try to define doppelganger detection periods here. Because we refactored doppelganger detection into a separate service we don't need enable it at this point anyways. When the service starts up (post-genesis) it'll always update doppelganger detection periods

paulhauner reviewed (changes requested):
Looking good, nice and tidy! I like the way this is heading.

I've left some comments, mostly just little things. Let me know your thoughts.

Something else that might be nice is a little page in the docs, explaining the feature and the costs associated with missing some attns versus getting slashed. LMK if you'd prefer me to write it.

This is a pretty serious feature so I'm being pretty detailed with the review. I still need to do another pass and check some other things. Here's a little checklist for myself:

- [ ] Try the bash script tests
- [ ] Try with some Pyrmont validators, see if I can get them slashed.

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
General PR updates

realbigsean committed:
Updates to match Paul's API spec proposal

realbigsean committed:
fix clippy errors

realbigsean commented:
Ok, ready for another round of review!

Here's a diff: https://github.com/realbigsean/lighthouse/compare/d136c08779444694519bd37f0432c77a5ead75ea..b700e85588be820a3eda83e43bea4d542d1659d3

paulhauner commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2815,6 +2815,22 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+    pub fn validator_seen_at_epoch(&self, validator_index: usize, epoch: &Epoch) -> bool {
+        self.observed_attesters
+            .read()
+            .index_seen_at_epoch(validator_index, epoch)
```
To avoid some obscure read-lock interleaving, I think we need to assign to intermediate vars and then compare those. Example:

```rust
let attested = self.observed_attesters
            .read()
            .index_seen_at_epoch(validator_index, epoch);
let aggregated = self
            .observed_aggregators
            .read()
            .index_seen_at_epoch(validator_index, epoch);
let produced_block = self
            .observed_block_producers
            .read()
            .index_seen_at_epoch(validator_index as u64, epoch);
            
attested || aggregated || produced_block
``` 

I made a [Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0c7657f4b14de9ac62eda7ef73e3bf31) to demonstrate. Here's the lock ordering without my recommendation:

```
without assigning to intermediate var:
locked a
locked b
unlocked b
locked c
unlocked c
unlocked a
```

We can see that we lock `b` and `c` whilst we're also holding the lock to `a`. If some other part of the program attempted to simultaneously do the reverse (hold `b` or `c` whilst trying to lock `a`) the program will deadlock.

Here's the lock ordering *with* my recommendation:

```
assigning to intermediate var:
locked a
unlocked a
locked b
unlocked b
locked c
unlocked c
```

I only know this because I've been bitten by it before, I had no idea before the first time I saw it. It's fairly counterintuitive IMO.

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1905,6 +1905,45 @@ pub fn serve<T: BeaconChainTypes>(
+        .and_then(
+            |request_data: api_types::LivenessRequestData, chain: Arc<BeaconChain<T>>| {
+                blocking_json_task(move || {
+                    let liveness: Vec<api_types::LivenessResponseData> = request_data
```
Perhaps we should add a check just before this to make sure the request epoch is within the prev/current/next epochs and return an error if it's not. This avoids giving false-negatives on old epochs.

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,141 @@
+
+                    if let Err(e) = doppelganger_service.detect_doppelgangers().await {
+                        error!(log,"Error during doppelganger detection"; "error" => ?e);
+                        break;
```
I think we should either:

1. Remove this `break`.
2. Call the shutdown signal, and then `break`.

Unless I'm missing something, currently if a call to `slot_clock` fails then we'll shutdown the doppleganger service without shutting down anything else.

I think we should go with (1) and just keep the service running. That way it can keep trying to make progress.

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1905,6 +1905,45 @@ pub fn serve<T: BeaconChainTypes>(
+                        .iter()
+                        .map(|id| {
+                            let index = match id {
+                                ValidatorId::PublicKey(pubkey) => chain
```
I can see that you've considered this in the rest of the code, but if we give a pubkey that's not in the head state then we'll invalidate the whole request.

Perhaps we're best to *only* accept indices here, I think that's how the API spec is written atm. That'll prevent us from getting into the place where we accidentally poison the request with an unknown validator pubkey.

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,141 @@
+            .read()
+            .get_doppelganger_detecting_validators(epoch);
+        let validators_slice = validators.as_slice();
+
```
It might be nice to add a `if validators_slice.is_empty() { return Ok(()) }` and avoid the empty request.

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -611,4 +657,81 @@ impl InitializedValidators {
+                                // We want to avoid checking the epoch in which doppelganger detection was started
+                                // so we don't pick up attestations from our own validator on restart.
+                                (doppelganger_epoch >= epoch
+                                    && epoch != doppelganger_epoch - DOPPELGANGER_DETECTION_EPOCHS)
```
Nitpick: I'm starting to go cold on relying on the saturating methods of slot/epoch since a seemingly innocent change of type to `u64` can cause unexpected consequences.

```suggestion
                                    && epoch != doppelganger_epoch.saturating_sub(DOPPELGANGER_DETECTION_EPOCHS))
```

paulhauner commented on validator_client/src/duties_service.rs:
```diff
@@ -357,16 +371,20 @@ async fn poll_beacon_attesters<T: SlotClock + 'static, E: EthSpec>(
 
     let local_indices = {
         let mut local_indices = Vec::with_capacity(local_pubkeys.len());
-        let indices_map = duties_service.indices.read();
```
What's the reasoning for switching to obtaining the lock on each iteration of `local_pubkeys`?

I'd be tempted to leave it there since it's safe (doesn't interact with other locks) and it'll avoid bouncing the lock when we have lots of validators. But, if you do want to remove it we can also remove the indenting where `local_indices` is assigned; it was only there to prevent the read-lock from "escaping" and living longer than we expected.

paulhauner commented on validator_client/src/initialized_validators.rs:
```diff
@@ -313,7 +333,7 @@ impl InitializedValidators {
     }
 
     /// Iterate through all **enabled** voting public keys in `self`.
```
```suggestion
    /// Iterate through all voting public keys in `self` that should be used when querying for duties.
```

paulhauner commented on validator_client/src/validator_store.rs:
```diff
@@ -113,17 +120,33 @@ impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
 
         Ok(validator_def)
     }
 
```
It would be nice to add some comments to these different pubkey-getting functions to describe why we have different sets of keys for duties and for signing.

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,141 @@
+            .beacon_nodes
+            .first_success(RequireSynced::Yes, |beacon_node| async move {
+                beacon_node
+                    .post_lighthouse_liveness(validators_slice, epoch)
```
With current setup of:

- Polling at the start of each slot
- Requesting the current epoch

I'm afraid that we're going to miss the case where a validator is assigned to the last slot in the epoch. In such a scenario we wouldn't observe them until ~1/3rd through the final slot (when the produce attestations) and doppleganger would have already completed its last poll for that epoch.

A potential solution here is to always poll from the current *and* previous epochs here. When we polled for the previous epoch we'd have to exclude validators that were instantiated in that epoch to avoid that thing where we detect our own attestations during a quick reboot.

paulhauner reviewed (changes requested):
Making progress! I have a few more comments after doing another thorough review.

One thing I've been considering is this case:

1. User starts their VC.
1. A few seconds later, the user shuts the lid of their laptop (i.e., suspends it).
1. User opens the lid and wakes the PC after several epochs have passed.

I'm not *certain* this is the case, but I think doppleganger protection will assume that all those epochs were checked successfully and start operating as per usual. At the very least, we're not guaranteed safety in this case.

A solution to this would be instead of having `def.doppleganger_detection_epoch` represent the epoch at which detection finishes, we could instead have `def.num_remaining_doppleganger_checks` and then decrease that after we've successfully checked an epoch. Once it hits `0` we know the validator is safe to run.

Checking determining when an epoch has been successfully checked is a little tricky since we check each slot, but doable I think. Perhaps we consider an epoch checked after we've performed the check of epoch `n - 1` (prev epoch) during the last slot of epoch `n` (current epoch).

Perhaps you've already considered this and I'm just covering old ground. Let me know what you think, happy to jump on call to chat if that's easier :)

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
address most of the new PR feedback

realbigsean committed:
Track completed epochs by decrementing `num_remaining_detection_epochs`

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	Cargo.lock

realbigsean committed:
cargo lock update

realbigsean commented:
> Perhaps we consider an epoch checked after we've performed the check of epoch n - 1 (prev epoch) during the last slot of epoch n (current epoch).

I implemented this as the considering an epoch checked after we've performed the check of epoch n - 1 (prev epoch) during the _first_ slot of epoch n.  It made more sense to me this way, let me know if you want me to update that

paulhauner committed:
Change break -> continue

paulhauner committed:
Change logs for global consistency

paulhauner committed:
Add incomplete progress

realbigsean committed:
sensitive url fix

paulhauner committed:
Restrict signing based on doppelganger

paulhauner committed:
Add comments, fix violator detection

maurodelazeri commented:
I don't think that this should be the default behavior... I think that the duplication of validators should be treated outside the validator node, in my opinion, it is outside the scope of what the validator should do.

dhigby commented:
I can't think of a reason how doppelganger detection would benefit a single validator operator like myself.  Having it turned on by default only makes sense If the majority of operators have multiple validators.  If not, let the validator farms activate it in their scripts and leave it turned off for us little guys.

paulhauner commented:
> I can't think of a reason how doppelganger detection would benefit a single validator operator like myself.

One scenario where the single-validator operator risks slashing is if they need to move their validator to a new computer. If they forget to turn the original validator off before they start the new one, then they're immediately in slashing territory that will (most likely) be protected by this feature.

If we start to consider users who might have a few validators across a couple machines, the possibilities start to increase. Then, it increases more once we consider users who might have some validators in a hold-your-own-keys staking pool (e.g., Allnodes).

> If not, let the validator farms activate it in their scripts and leave it turned off for us little guys.

In my experience, the staking pools are running rather complicated setups which wouldn't really benefit from this. This feature is primarily intended as a last-line-of-defense for "little guys/gals".

---

It's important to note that one validator missing 4 attestations costs approximately 30 cents (USD) at current prices.

The ethos of Lighthouse so far has very much been "safety by default" and I'm not convinced that we should break that to save less than a dollar each time the VC restarts.

I understand that some users have a particular attachment to getting perfect attestation scores, but I'm also aware of a much stronger, community-wide fear of getting slashed. Considering these competing fears and the low cost of safety-by-default, I'm finding it difficult to make an argument to disabling this by default. Savvy users can use the flag to disable the feature whilst the naive are protected by default.

If there are arguments as to why this might be *structurally* bad for the Beacon Chain, then I'd love to hear about it. Presently, I can see it as structurally *good* since every slashed validator bloats the `BeaconState` and adds to the collective fear of staking. 

A reason why it might be structurally bad is if there's some vuln that causes a validator client to reboot. Then, the chain will experience a couple more epochs downtime. Even in this scenario, I'm not sure that a couple of more epochs downtime are *that bad*, especially once we consider that there's going to be a bunch of power users opting out of this feature (I will most likely opt-out myself and I suspect staking pools will too).

paulhauner committed:
Merge branch 'unstable' into sean-dopple

paulhauner committed:
Add comment

paulhauner committed:
Further tidying

paulhauner committed:
Introduce VotingPubkey enum

paulhauner committed:
Rename function

paulhauner committed:
Fix test compile error

paulhauner committed:
Tidy

paulhauner committed:
Rename VotingPubkey -> DoppelgangerStatus

paulhauner committed:
Add doppelganger protection comment

paulhauner committed:
Rename DoppelgangerStatus fns

paulhauner committed:
Tidy comments

paulhauner committed:
Skip validators with inconsistencies, tidy

paulhauner committed:
Don't exit if BN returns nothing for liveness

paulhauner committed:
Appease clippy

paulhauner committed:
Refactor to smaller functions

paulhauner committed:
Make `detect_doppelgangers` more generic

paulhauner committed:
Use generic shutdown signal

paulhauner committed:
Add first tests

paulhauner committed:
Add more testing

paulhauner committed:
More testing

paulhauner committed:
Add success scenario test

paulhauner committed:
Add skip forward test

paulhauner committed:
Add time skip backwards test

paulhauner committed:
Add staggered entry test

paulhauner committed:
Remove slot clock from doppelganger

paulhauner committed:
Update comment

paulhauner committed:
Allow selection proof signing

paulhauner committed:
Add comment

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	scripts/local_testnet/ganache_test_node.sh
	scripts/local_testnet/setup.sh
	validator_client/src/cli.rs
	validator_client/src/config.rs

realbigsean committed:
Fixes related to merge with `unstable`

realbigsean committed:
add newline

realbigsean committed:
test shutdown reason success exit code

realbigsean committed:
Test failure

realbigsean committed:
rust 1.53.0 updates

realbigsean committed:
add recursion limit to `simulator` crate

realbigsean committed:
cargo fmt

realbigsean committed:
Merge branch 'rust-1.53.0-lints' of https://github.com/realbigsean/lighthouse into doppleganger-detection

realbigsean committed:
Test success

realbigsean committed:
Merge branch 'doppleganger-detection' of https://github.com/realbigsean/lighthouse into paul-sean-dopple

 Conflicts:
	scripts/local_testnet/ganache_test_node.sh
	scripts/local_testnet/setup.sh
	validator_client/src/doppelganger_service.rs

realbigsean committed:
fix duplicate `Cargo.toml` entry

realbigsean committed:
- move `DetectionState` mutation to method.
- calculate next epoch in `beacon_node_liveness` method

realbigsean committed:
fix comments, set `DEFAULT_REMAINING_DETECTION_EPOCHS` to 1

realbigsean committed:
Disable doppelganger detection by default

realbigsean committed:
Add doppelganger status endpoint

realbigsean committed:
Add doppelganger protection enabled method

realbigsean committed:
Add docs, fix tests and some comments

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	beacon_node/eth2_libp2p/src/peer_manager/peer_sync_status.rs

realbigsean committed:
Merge branch 'doppleganger-detection' of https://github.com/realbigsean/lighthouse into paul-sean-dopple

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into paul-sean-dopple

 Conflicts:
	Cargo.lock

realbigsean committed:
- And an `Arc` around the `ValidatorStore`
- Remove `Option` from `validator_store.doppelganger_detection`
- rename doppelganger detection to doppelganger protection

realbigsean committed:
- Add local testnet test for successful doppelganger protection period
- Update logging during doppelganger protection period

realbigsean committed:
remove unncessary `mut`

realbigsean committed:
update doppelganger test in ci

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	Cargo.lock
	beacon_node/beacon_chain/src/observed_block_producers.rs
	scripts/local_testnet/vars.env
	validator_client/src/http_api/mod.rs

realbigsean committed:
update cargo lock

realbigsean committed:
Merge branch 'doppleganger-detection' of https://github.com/realbigsean/lighthouse into paul-sean-dopple

 Conflicts:
	validator_client/src/http_api/mod.rs

realbigsean committed:
Add a timeout for the liveness request

realbigsean committed:
fix test compilation after merge

paulhauner committed:
Use Option instead of enabled bool

realbigsean committed:
fix typo and add fork slot to test `.env` files

realbigsean committed:
fix tests

realbigsean committed:
Merge pull request #16 from realbigsean/paul-sean-dopple

Paul's Doppelganger Protection Updates

WALL-E commented:
great !

realbigsean committed:
improve doppelganger success test

realbigsean committed:
update seconds per slot on success

realbigsean committed:
fix local testnet

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	beacon_node/beacon_chain/src/observed_attesters.rs

realbigsean committed:
self review updates

realbigsean committed:
Revert "Add doppelganger status endpoint"

realbigsean committed:
Small fixes

realbigsean committed:
make seconds per eth2 slot default to 3 in scripts

realbigsean committed:
remove duplicate vars.env file

realbigsean commented:
I previously added an endpoint for the UI to see which validators are in a doppelganger detection period, but removed it from this PR (https://github.com/sigp/lighthouse/pull/2230/commits/26a6c0dfb2f56f4a4fe2e1040d2745ad35b72ef1) because I think it needs some improvements and isn't necessary for the doppelganger protection process. I'm going to work on it in a separate PR.

realbigsean committed:
add cli test for doppelganger flag

realbigsean committed:
verbose commands

realbigsean committed:
add lines to debug

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
update `is_current_epoch_attester` to `is_current_epoch_target_attester` in doppelganger test script

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

realbigsean committed:
increase bootnode startup wait time

realbigsean committed:
increase seconds per slot

realbigsean committed:
increase seconds per slot

realbigsean committed:
increase seconds per slot

realbigsean committed:
update to use previous  epoch inclusion

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1909,6 +1909,49 @@ pub fn serve<T: BeaconChainTypes>(
+                        .map(|index| {
+                            let is_live =
+                                chain.validator_seen_at_epoch(index as usize, request_data.epoch);
+                            Ok::<_, warp::Rejection>(api_types::LivenessResponseData {
```
If we remove this `Result` the iterator will become fixed-size and should therefore require less allocations. I'm pretty sure we don't need the `Result` there.

paulhauner commented on lighthouse/tests/validator_client.rs:
```diff
@@ -378,3 +378,10 @@ pub fn malloc_tuning_flag() {
         .run();
 }
+#[test]
+fn doppelganger_protection_flag() {
```
It might be nice to also add the inverse of this check; that doppelganger is *disabled* by default. That would help prevent us accidentally turning it on :grimacing:

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,1325 @@
+//!
+//! The mechanism works roughly like so: when the validator client starts or a new validator is
+//! added, that validator is assigned a number of "remaining epochs". The doppelganger service
+//! periodically poll the beacon node to if that validator has been observed to produce
```
```suggestion
//! periodically polls the beacon node to see if that validator has been observed to produce
```

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,1325 @@
+pub enum DoppelgangerStatus {
+    /// Doppelganger protection has approved this for signing.
+    ///
+    /// This is because the service is disabled, or the service has waited some period of time to
```
```suggestion
    /// This is because the service has waited some period of time to
```

paulhauner commented on validator_client/src/doppelganger_service.rs:
```diff
@@ -0,0 +1,1325 @@
+    }
+
+    /// Returns a key regardless of whether or not doppelganger has approved it. Such a key might be
+    /// used for signing, duties collection or other activities.
```
```suggestion
    /// used for signing non-slashable messages, duties collection or other activities.
```

paulhauner reviewed (changes requested):
Great work on the tests! We're sooooo close.

I just have a couple of really minor things and corrections to *my own* typos.

:tada:

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into doppleganger-detection

 Conflicts:
	validator_client/src/duties_service.rs

realbigsean committed:
Update validator_client/src/doppelganger_service.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Update validator_client/src/doppelganger_service.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Update validator_client/src/doppelganger_service.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Merge branch 'doppleganger-detection' of https://github.com/realbigsean/lighthouse into doppleganger-detection

realbigsean committed:
PR updates plus a few extra lints

paulhauner reviewed (approved):
Perfect! Merge time :tada: 

bors r+

michaelsproul commented:
Well done @realbigsean @paulhauner 👏 🎉
